vcd/resource_vcd_service_account.go

```diff
@@ -0,0 +1,349 @@
+		SoftwareVersion: softwareVersion,
+		URI:             uri,
+		Role: &types.OpenApiReference{
+			ID: roleId,
+		},
+	}
+
+	_, err = sa.Update(saConfig)
+	if err != nil {
+		return diag.Errorf("[Service Account update] error updating Service Account: %s", err)
+	}
+
+	return resourceVcdServiceAccountRead(ctx, d, meta)
+}
+
+func updateServiceAccountStatus(sa *govcd.ServiceAccount, active bool, filename, useragent string) error {
+	if active {
+		err := sa.Authorize()
+		if err != nil {
+			return fmt.Errorf("error authorizing Service Account: %s", err)
+		}
+		err = sa.Refresh()
+		if err != nil {
+			return fmt.Errorf("error refreshing Service Account: %s", err)
+		}
+		err = sa.Grant()
+		if err != nil {
+			return fmt.Errorf("error granting Service Account: %s", err)
+		}
+		err = sa.Refresh()
+		if err != nil {
+			return fmt.Errorf("error refreshing Service Account: %s", err)
+		}
+		initialApiToken, err := sa.GetInitialApiToken()
+		if err != nil {
+			return fmt.Errorf("error activating Service Account: %s", err)
+		}
+		err = sa.Refresh()
+		if err != nil {
+			return fmt.Errorf("error refreshing Service Account: %s", err)
+		}
+		err = govcd.SaveServiceAccountToFile(filename, useragent, initialApiToken)
+		if err != nil {
+			return fmt.Errorf("error saving service account api token to file: %s", err)
+		}
+	} else {
+		err := sa.Revoke()
+		if err != nil {
+			return fmt.Errorf("error revoking Service Account: %s", err)
+		}
+		err = sa.Refresh()
+		if err != nil {
+			return fmt.Errorf("error refreshing Service Account: %s", err)
+		}
+	}
+
+	return nil
+}
+
+func resourceVcdServiceAccountRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	return genericVcdServiceAccountRead(ctx, d, meta, "resource")
+}
+
+func genericVcdServiceAccountRead(ctx context.Context, d *schema.ResourceData, meta interface{}, origin string) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	org, err := vcdClient.GetOrgFromResource(d)
+	if err != nil {
+		return diag.Errorf(errorRetrievingOrg, err)
+	}
+	var sa *govcd.ServiceAccount
+	if d.Id() != "" {
+		sa, err = org.GetServiceAccountById(d.Id())
+	} else {
+		saName := d.Get("name").(string)
+		sa, err = org.GetServiceAccountByName(saName)
+	}
+	if govcd.ContainsNotFound(err) {
+		// When parent Edge Gateway is not found - this resource is also not found and should be
+		// removed from state
+		if origin == "datasource" {
+			return diag.Errorf("[Service Account DS read] error retrieving Service Account: %s", err)
+		}
+		d.SetId("")
+		log.Printf("[DEBUG] Service Account no longer exists. Removing from tfstate")
+		return nil
+	}
+
+	if err != nil {
+		return diag.Errorf("[Service Account read] error: %s", err)
+	}
+
+	d.SetId(sa.ServiceAccount.ID)
+	dSet(d, "name", sa.ServiceAccount.Name)
+	dSet(d, "software_id", sa.ServiceAccount.SoftwareID)
+	dSet(d, "software_version", sa.ServiceAccount.SoftwareVersion)
+	dSet(d, "role_id", sa.ServiceAccount.Role.ID)
```

Comment:
Just double-checking: Any chances of `Role *OpenApiReference` being `nil`?
(Same for line 343/Import operation, just to not to repeat)

Reply:
I tried to check in many ways, but Role should never be nil, it's one of the pre-requisites for service accounts.